File: LC/[0376] Wiggle Subsequence -  DP.cpp
class Solution {
public:
    int wiggleMaxLength(vector<int>& nums) {
        if(nums.size()==0) return 0;
        
        /*
        let p be max wiggle len ending @ down-slope
            q be max wiggle len ending @ up-slope
        */
        int p=1, q=1;
        for(int ii=1; ii<nums.size(); ++ii){
            if(nums[ii] > nums[ii-1]){//current is uplope
                q=p+1;//upslope can increment down-slope
            }else if(nums[ii] < nums[ii-1]){//current is down-slope
                p=q+1;//down-slope can only increment upslope
            }
        }

        return max(p, q);
    }
};

File: LC/[0376] Wiggle Subsequence - dp.py
class Solution:
    def wiggleMaxLength(self, nums: List[int]) -> int:
        up=1 #dp max wiggle len ending @ up slope
        dn=1 #dp max wiggle len ending @ down slope
        N=len(nums)
        for ii in range(1, N):
            if nums[ii] > nums[ii-1]: up = dn + 1
            if nums[ii] < nums[ii-1]: dn = up +1
        
        return max(up, dn)